File: android/src/main/java/com/vnptmap/sdk/v2/flutter_vnptmap_gl/LayerPropertyConverter.java
package com.vnptmap.sdk.v2.flutter_vnptmap_gl;

import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

import com.mapbox.mapboxsdk.style.expressions.Expression;
import com.mapbox.mapboxsdk.style.layers.PropertyFactory;
import com.mapbox.mapboxsdk.style.layers.PropertyValue;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import static com.vnptmap.sdk.v2.flutter_vnptmap_gl.VNPTMapConvert.toMap;


public class LayerPropertyConverter {
    static PropertyValue[] interpretSymbolLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "icon-opacity":
                    properties.add(PropertyFactory.iconOpacity(expression));
                    break;
                case "icon-color":
                    properties.add(PropertyFactory.iconColor(expression));
                    break;
                case "icon-halo-color":
                    properties.add(PropertyFactory.iconHaloColor(expression));
                    break;
                case "icon-halo-width":
                    properties.add(PropertyFactory.iconHaloWidth(expression));
                    break;
                case "icon-halo-blur":
                    properties.add(PropertyFactory.iconHaloBlur(expression));
                    break;
                case "icon-translate":
                    properties.add(PropertyFactory.iconTranslate(expression));
                    break;
                case "icon-translate-anchor":
                    properties.add(PropertyFactory.iconTranslateAnchor(expression));
                    break;
                case "text-opacity":
                    properties.add(PropertyFactory.textOpacity(expression));
                    break;
                case "text-color":
                    properties.add(PropertyFactory.textColor(expression));
                    break;
                case "text-halo-color":
                    properties.add(PropertyFactory.textHaloColor(expression));
                    break;
                case "text-halo-width":
                    properties.add(PropertyFactory.textHaloWidth(expression));
                    break;
                case "text-halo-blur":
                    properties.add(PropertyFactory.textHaloBlur(expression));
                    break;
                case "text-translate":
                    properties.add(PropertyFactory.textTranslate(expression));
                    break;
                case "text-translate-anchor":
                    properties.add(PropertyFactory.textTranslateAnchor(expression));
                    break;
                case "symbol-placement":
                    properties.add(PropertyFactory.symbolPlacement(expression));
                    break;
                case "symbol-spacing":
                    properties.add(PropertyFactory.symbolSpacing(expression));
                    break;
                case "symbol-avoid-edges":
                    properties.add(PropertyFactory.symbolAvoidEdges(expression));
                    break;
                case "symbol-sort-key":
                    properties.add(PropertyFactory.symbolSortKey(expression));
                    break;
                case "symbol-z-order":
                    properties.add(PropertyFactory.symbolZOrder(expression));
                    break;
                case "icon-allow-overlap":
                    properties.add(PropertyFactory.iconAllowOverlap(expression));
                    break;
                case "icon-ignore-placement":
                    properties.add(PropertyFactory.iconIgnorePlacement(expression));
                    break;
                case "icon-optional":
                    properties.add(PropertyFactory.iconOptional(expression));
                    break;
                case "icon-rotation-alignment":
                    properties.add(PropertyFactory.iconRotationAlignment(expression));
                    break;
                case "icon-size":
                    properties.add(PropertyFactory.iconSize(expression));
                    break;
                case "icon-text-fit":
                    properties.add(PropertyFactory.iconTextFit(expression));
                    break;
                case "icon-text-fit-padding":
                    properties.add(PropertyFactory.iconTextFitPadding(expression));
                    break;
                case "icon-image":
                    if (jsonElement.isJsonPrimitive() && jsonElement.getAsJsonPrimitive().isString()) {
                        properties.add(PropertyFactory.iconImage(jsonElement.getAsString()));
                    } else {
                        properties.add(PropertyFactory.iconImage(expression));
                    }
                    break;
                case "icon-rotate":
                    properties.add(PropertyFactory.iconRotate(expression));
                    break;
                case "icon-padding":
                    properties.add(PropertyFactory.iconPadding(expression));
                    break;
                case "icon-keep-upright":
                    properties.add(PropertyFactory.iconKeepUpright(expression));
                    break;
                case "icon-offset":
                    properties.add(PropertyFactory.iconOffset(expression));
                    break;
                case "icon-anchor":
                    properties.add(PropertyFactory.iconAnchor(expression));
                    break;
                case "icon-pitch-alignment":
                    properties.add(PropertyFactory.iconPitchAlignment(expression));
                    break;
                case "text-pitch-alignment":
                    properties.add(PropertyFactory.textPitchAlignment(expression));
                    break;
                case "text-rotation-alignment":
                    properties.add(PropertyFactory.textRotationAlignment(expression));
                    break;
                case "text-field":
                    properties.add(PropertyFactory.textField(expression));
                    break;
                case "text-font":
                    properties.add(PropertyFactory.textFont(expression));
                    break;
                case "text-size":
                    properties.add(PropertyFactory.textSize(expression));
                    break;
                case "text-max-width":
                    properties.add(PropertyFactory.textMaxWidth(expression));
                    break;
                case "text-line-height":
                    properties.add(PropertyFactory.textLineHeight(expression));
                    break;
                case "text-letter-spacing":
                    properties.add(PropertyFactory.textLetterSpacing(expression));
                    break;
                case "text-justify":
                    properties.add(PropertyFactory.textJustify(expression));
                    break;
                case "text-radial-offset":
                    properties.add(PropertyFactory.textRadialOffset(expression));
                    break;
                case "text-variable-anchor":
                    properties.add(PropertyFactory.textVariableAnchor(expression));
                    break;
                case "text-anchor":
                    properties.add(PropertyFactory.textAnchor(expression));
                    break;
                case "text-max-angle":
                    properties.add(PropertyFactory.textMaxAngle(expression));
                    break;
                case "text-writing-mode":
                    properties.add(PropertyFactory.textWritingMode(expression));
                    break;
                case "text-rotate":
                    properties.add(PropertyFactory.textRotate(expression));
                    break;
                case "text-padding":
                    properties.add(PropertyFactory.textPadding(expression));
                    break;
                case "text-keep-upright":
                    properties.add(PropertyFactory.textKeepUpright(expression));
                    break;
                case "text-transform":
                    properties.add(PropertyFactory.textTransform(expression));
                    break;
                case "text-offset":
                    properties.add(PropertyFactory.textOffset(expression));
                    break;
                case "text-allow-overlap":
                    properties.add(PropertyFactory.textAllowOverlap(expression));
                    break;
                case "text-ignore-placement":
                    properties.add(PropertyFactory.textIgnorePlacement(expression));
                    break;
                case "text-optional":
                    properties.add(PropertyFactory.textOptional(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretCircleLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "circle-radius":
                    properties.add(PropertyFactory.circleRadius(expression));
                    break;
                case "circle-color":
                    properties.add(PropertyFactory.circleColor(expression));
                    break;
                case "circle-blur":
                    properties.add(PropertyFactory.circleBlur(expression));
                    break;
                case "circle-opacity":
                    properties.add(PropertyFactory.circleOpacity(expression));
                    break;
                case "circle-translate":
                    properties.add(PropertyFactory.circleTranslate(expression));
                    break;
                case "circle-translate-anchor":
                    properties.add(PropertyFactory.circleTranslateAnchor(expression));
                    break;
                case "circle-pitch-scale":
                    properties.add(PropertyFactory.circlePitchScale(expression));
                    break;
                case "circle-pitch-alignment":
                    properties.add(PropertyFactory.circlePitchAlignment(expression));
                    break;
                case "circle-stroke-width":
                    properties.add(PropertyFactory.circleStrokeWidth(expression));
                    break;
                case "circle-stroke-color":
                    properties.add(PropertyFactory.circleStrokeColor(expression));
                    break;
                case "circle-stroke-opacity":
                    properties.add(PropertyFactory.circleStrokeOpacity(expression));
                    break;
                case "circle-sort-key":
                    properties.add(PropertyFactory.circleSortKey(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretLineLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "line-opacity":
                    properties.add(PropertyFactory.lineOpacity(expression));
                    break;
                case "line-color":
                    properties.add(PropertyFactory.lineColor(expression));
                    break;
                case "line-translate":
                    properties.add(PropertyFactory.lineTranslate(expression));
                    break;
                case "line-translate-anchor":
                    properties.add(PropertyFactory.lineTranslateAnchor(expression));
                    break;
                case "line-width":
                    properties.add(PropertyFactory.lineWidth(expression));
                    break;
                case "line-gap-width":
                    properties.add(PropertyFactory.lineGapWidth(expression));
                    break;
                case "line-offset":
                    properties.add(PropertyFactory.lineOffset(expression));
                    break;
                case "line-blur":
                    properties.add(PropertyFactory.lineBlur(expression));
                    break;
                case "line-dasharray":
                    properties.add(PropertyFactory.lineDasharray(expression));
                    break;
                case "line-pattern":
                    properties.add(PropertyFactory.linePattern(expression));
                    break;
                case "line-gradient":
                    properties.add(PropertyFactory.lineGradient(expression));
                    break;
                case "line-cap":
                    properties.add(PropertyFactory.lineCap(expression));
                    break;
                case "line-join":
                    properties.add(PropertyFactory.lineJoin(expression));
                    break;
                case "line-miter-limit":
                    properties.add(PropertyFactory.lineMiterLimit(expression));
                    break;
                case "line-round-limit":
                    properties.add(PropertyFactory.lineRoundLimit(expression));
                    break;
                case "line-sort-key":
                    properties.add(PropertyFactory.lineSortKey(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretFillLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "fill-antialias":
                    properties.add(PropertyFactory.fillAntialias(expression));
                    break;
                case "fill-opacity":
                    properties.add(PropertyFactory.fillOpacity(expression));
                    break;
                case "fill-color":
                    properties.add(PropertyFactory.fillColor(expression));
                    break;
                case "fill-outline-color":
                    properties.add(PropertyFactory.fillOutlineColor(expression));
                    break;
                case "fill-translate":
                    properties.add(PropertyFactory.fillTranslate(expression));
                    break;
                case "fill-translate-anchor":
                    properties.add(PropertyFactory.fillTranslateAnchor(expression));
                    break;
                case "fill-pattern":
                    properties.add(PropertyFactory.fillPattern(expression));
                    break;
                case "fill-sort-key":
                    properties.add(PropertyFactory.fillSortKey(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretFillExtrusionLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "fill-extrusion-opacity":
                    properties.add(PropertyFactory.fillExtrusionOpacity(expression));
                    break;
                case "fill-extrusion-color":
                    properties.add(PropertyFactory.fillExtrusionColor(expression));
                    break;
                case "fill-extrusion-translate":
                    properties.add(PropertyFactory.fillExtrusionTranslate(expression));
                    break;
                case "fill-extrusion-translate-anchor":
                    properties.add(PropertyFactory.fillExtrusionTranslateAnchor(expression));
                    break;
                case "fill-extrusion-pattern":
                    properties.add(PropertyFactory.fillExtrusionPattern(expression));
                    break;
                case "fill-extrusion-height":
                    properties.add(PropertyFactory.fillExtrusionHeight(expression));
                    break;
                case "fill-extrusion-base":
                    properties.add(PropertyFactory.fillExtrusionBase(expression));
                    break;
                case "fill-extrusion-vertical-gradient":
                    properties.add(PropertyFactory.fillExtrusionVerticalGradient(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretRasterLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "raster-opacity":
                    properties.add(PropertyFactory.rasterOpacity(expression));
                    break;
                case "raster-hue-rotate":
                    properties.add(PropertyFactory.rasterHueRotate(expression));
                    break;
                case "raster-brightness-min":
                    properties.add(PropertyFactory.rasterBrightnessMin(expression));
                    break;
                case "raster-brightness-max":
                    properties.add(PropertyFactory.rasterBrightnessMax(expression));
                    break;
                case "raster-saturation":
                    properties.add(PropertyFactory.rasterSaturation(expression));
                    break;
                case "raster-contrast":
                    properties.add(PropertyFactory.rasterContrast(expression));
                    break;
                case "raster-resampling":
                    properties.add(PropertyFactory.rasterResampling(expression));
                    break;
                case "raster-fade-duration":
                    properties.add(PropertyFactory.rasterFadeDuration(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretHillshadeLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "hillshade-illumination-direction":
                    properties.add(PropertyFactory.hillshadeIlluminationDirection(expression));
                    break;
                case "hillshade-illumination-anchor":
                    properties.add(PropertyFactory.hillshadeIlluminationAnchor(expression));
                    break;
                case "hillshade-exaggeration":
                    properties.add(PropertyFactory.hillshadeExaggeration(expression));
                    break;
                case "hillshade-shadow-color":
                    properties.add(PropertyFactory.hillshadeShadowColor(expression));
                    break;
                case "hillshade-highlight-color":
                    properties.add(PropertyFactory.hillshadeHighlightColor(expression));
                    break;
                case "hillshade-accent-color":
                    properties.add(PropertyFactory.hillshadeAccentColor(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }

    static PropertyValue[] interpretHeatmapLayerProperties(Object o) {
        final Map<String, String> data = (Map<String, String>) toMap(o);
        final List<PropertyValue> properties = new LinkedList();
        final JsonParser parser = new JsonParser();

        for (Map.Entry<String, String> entry : data.entrySet()) {
            final JsonElement jsonElement = parser.parse(entry.getValue());
            Expression expression = Expression.Converter.convert(jsonElement);
            switch (entry.getKey()) {
                case "heatmap-radius":
                    properties.add(PropertyFactory.heatmapRadius(expression));
                    break;
                case "heatmap-weight":
                    properties.add(PropertyFactory.heatmapWeight(expression));
                    break;
                case "heatmap-intensity":
                    properties.add(PropertyFactory.heatmapIntensity(expression));
                    break;
                case "heatmap-color":
                    properties.add(PropertyFactory.heatmapColor(expression));
                    break;
                case "heatmap-opacity":
                    properties.add(PropertyFactory.heatmapOpacity(expression));
                    break;
                case "visibility":
                    properties.add(PropertyFactory.visibility(entry.getValue()));
                    break;
                default:
                    break;
            }
        }

        return properties.toArray(new PropertyValue[properties.size()]);
    }
}

File: ios/Classes/LayerPropertyConverter.swift
//
//  LayerPropertiesConverter.swift
//  flutter_vnptmap_gl
//
//  Created by Võ Toàn on 13/12/2022.
//

import Mapbox

class LayerPropertyConverter {
    class func addSymbolProperties(symbolLayer: MGLSymbolStyleLayer, properties: [String: String]) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "icon-opacity":
                symbolLayer.iconOpacity = expression
            case "icon-color":
                symbolLayer.iconColor = expression
            case "icon-halo-color":
                symbolLayer.iconHaloColor = expression
            case "icon-halo-width":
                symbolLayer.iconHaloWidth = expression
            case "icon-halo-blur":
                symbolLayer.iconHaloBlur = expression
            case "icon-translate":
                symbolLayer.iconTranslation = expression
            case "icon-translate-anchor":
                symbolLayer.iconTranslationAnchor = expression
            case "text-opacity":
                symbolLayer.textOpacity = expression
            case "text-color":
                symbolLayer.textColor = expression
            case "text-halo-color":
                symbolLayer.textHaloColor = expression
            case "text-halo-width":
                symbolLayer.textHaloWidth = expression
            case "text-halo-blur":
                symbolLayer.textHaloBlur = expression
            case "text-translate":
                symbolLayer.textTranslation = expression
            case "text-translate-anchor":
                symbolLayer.textTranslationAnchor = expression
            case "symbol-placement":
                symbolLayer.symbolPlacement = expression
            case "symbol-spacing":
                symbolLayer.symbolSpacing = expression
            case "symbol-avoid-edges":
                symbolLayer.symbolAvoidsEdges = expression
            case "symbol-sort-key":
                symbolLayer.symbolSortKey = expression
            case "symbol-z-order":
                symbolLayer.symbolZOrder = expression
            case "icon-allow-overlap":
                symbolLayer.iconAllowsOverlap = expression
            case "icon-ignore-placement":
                symbolLayer.iconIgnoresPlacement = expression
            case "icon-optional":
                symbolLayer.iconOptional = expression
            case "icon-rotation-alignment":
                symbolLayer.iconRotationAlignment = expression
            case "icon-size":
                symbolLayer.iconScale = expression
            case "icon-text-fit":
                symbolLayer.iconTextFit = expression
            case "icon-text-fit-padding":
                symbolLayer.iconTextFitPadding = expression
            case "icon-image":
                symbolLayer.iconImageName = expression
            case "icon-rotate":
                symbolLayer.iconRotation = expression
            case "icon-padding":
                symbolLayer.iconPadding = expression
            case "icon-keep-upright":
                symbolLayer.keepsIconUpright = expression
            case "icon-offset":
                symbolLayer.iconOffset = expression
            case "icon-anchor":
                symbolLayer.iconAnchor = expression
            case "icon-pitch-alignment":
                symbolLayer.iconPitchAlignment = expression
            case "text-pitch-alignment":
                symbolLayer.textPitchAlignment = expression
            case "text-rotation-alignment":
                symbolLayer.textRotationAlignment = expression
            case "text-field":
                symbolLayer.text = expression
            case "text-font":
                symbolLayer.textFontNames = expression
            case "text-size":
                symbolLayer.textFontSize = expression
            case "text-max-width":
                symbolLayer.maximumTextWidth = expression
            case "text-line-height":
                symbolLayer.textLineHeight = expression
            case "text-letter-spacing":
                symbolLayer.textLetterSpacing = expression
            case "text-justify":
                symbolLayer.textJustification = expression
            case "text-radial-offset":
                symbolLayer.textRadialOffset = expression
            case "text-variable-anchor":
                symbolLayer.textVariableAnchor = expression
            case "text-anchor":
                symbolLayer.textAnchor = expression
            case "text-max-angle":
                symbolLayer.maximumTextAngle = expression
            case "text-writing-mode":
                symbolLayer.textWritingModes = expression
            case "text-rotate":
                symbolLayer.textRotation = expression
            case "text-padding":
                symbolLayer.textPadding = expression
            case "text-keep-upright":
                symbolLayer.keepsTextUpright = expression
            case "text-transform":
                symbolLayer.textTransform = expression
            case "text-offset":
                symbolLayer.textOffset = expression
            case "text-allow-overlap":
                symbolLayer.textAllowsOverlap = expression
            case "text-ignore-placement":
                symbolLayer.textIgnoresPlacement = expression
            case "text-optional":
                symbolLayer.textOptional = expression
            case "visibility":
                symbolLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addCircleProperties(circleLayer: MGLCircleStyleLayer, properties: [String: String]) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "circle-radius":
                circleLayer.circleRadius = expression
            case "circle-color":
                circleLayer.circleColor = expression
            case "circle-blur":
                circleLayer.circleBlur = expression
            case "circle-opacity":
                circleLayer.circleOpacity = expression
            case "circle-translate":
                circleLayer.circleTranslation = expression
            case "circle-translate-anchor":
                circleLayer.circleTranslationAnchor = expression
            case "circle-pitch-scale":
                circleLayer.circleScaleAlignment = expression
            case "circle-pitch-alignment":
                circleLayer.circlePitchAlignment = expression
            case "circle-stroke-width":
                circleLayer.circleStrokeWidth = expression
            case "circle-stroke-color":
                circleLayer.circleStrokeColor = expression
            case "circle-stroke-opacity":
                circleLayer.circleStrokeOpacity = expression
            case "circle-sort-key":
                circleLayer.circleSortKey = expression
            case "visibility":
                circleLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addLineProperties(lineLayer: MGLLineStyleLayer, properties: [String: String]) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "line-opacity":
                lineLayer.lineOpacity = expression
            case "line-color":
                lineLayer.lineColor = expression
            case "line-translate":
                lineLayer.lineTranslation = expression
            case "line-translate-anchor":
                lineLayer.lineTranslationAnchor = expression
            case "line-width":
                lineLayer.lineWidth = expression
            case "line-gap-width":
                lineLayer.lineGapWidth = expression
            case "line-offset":
                lineLayer.lineOffset = expression
            case "line-blur":
                lineLayer.lineBlur = expression
            case "line-dasharray":
                lineLayer.lineDashPattern = expression
            case "line-pattern":
                lineLayer.linePattern = expression
            case "line-gradient":
                lineLayer.lineGradient = expression
            case "line-cap":
                lineLayer.lineCap = expression
            case "line-join":
                lineLayer.lineJoin = expression
            case "line-miter-limit":
                lineLayer.lineMiterLimit = expression
            case "line-round-limit":
                lineLayer.lineRoundLimit = expression
            case "line-sort-key":
                lineLayer.lineSortKey = expression
            case "visibility":
                lineLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addFillProperties(fillLayer: MGLFillStyleLayer, properties: [String: String]) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "fill-antialias":
                fillLayer.fillAntialiased = expression
            case "fill-opacity":
                fillLayer.fillOpacity = expression
            case "fill-color":
                fillLayer.fillColor = expression
            case "fill-outline-color":
                fillLayer.fillOutlineColor = expression
            case "fill-translate":
                fillLayer.fillTranslation = expression
            case "fill-translate-anchor":
                fillLayer.fillTranslationAnchor = expression
            case "fill-pattern":
                fillLayer.fillPattern = expression
            case "fill-sort-key":
                fillLayer.fillSortKey = expression
            case "visibility":
                fillLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addFillExtrusionProperties(
        fillExtrusionLayer: MGLFillExtrusionStyleLayer,
        properties: [String: String]
    ) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "fill-extrusion-opacity":
                fillExtrusionLayer.fillExtrusionOpacity = expression
            case "fill-extrusion-color":
                fillExtrusionLayer.fillExtrusionColor = expression
            case "fill-extrusion-translate":
                fillExtrusionLayer.fillExtrusionTranslation = expression
            case "fill-extrusion-translate-anchor":
                fillExtrusionLayer.fillExtrusionTranslationAnchor = expression
            case "fill-extrusion-pattern":
                fillExtrusionLayer.fillExtrusionPattern = expression
            case "fill-extrusion-height":
                fillExtrusionLayer.fillExtrusionHeight = expression
            case "fill-extrusion-base":
                fillExtrusionLayer.fillExtrusionBase = expression
            case "fill-extrusion-vertical-gradient":
                fillExtrusionLayer.fillExtrusionHasVerticalGradient = expression
            case "visibility":
                fillExtrusionLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addRasterProperties(rasterLayer: MGLRasterStyleLayer, properties: [String: String]) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "raster-opacity":
                rasterLayer.rasterOpacity = expression
            case "raster-hue-rotate":
                rasterLayer.rasterHueRotation = expression
            case "raster-brightness-min":
                rasterLayer.minimumRasterBrightness = expression
            case "raster-brightness-max":
                rasterLayer.maximumRasterBrightness = expression
            case "raster-saturation":
                rasterLayer.rasterSaturation = expression
            case "raster-contrast":
                rasterLayer.rasterContrast = expression
            case "raster-resampling":
                rasterLayer.rasterResamplingMode = expression
            case "raster-fade-duration":
                rasterLayer.rasterFadeDuration = expression
            case "visibility":
                rasterLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addHillshadeProperties(
        hillshadeLayer: MGLHillshadeStyleLayer,
        properties: [String: String]
    ) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "hillshade-illumination-direction":
                hillshadeLayer.hillshadeIlluminationDirection = expression
            case "hillshade-illumination-anchor":
                hillshadeLayer.hillshadeIlluminationAnchor = expression
            case "hillshade-exaggeration":
                hillshadeLayer.hillshadeExaggeration = expression
            case "hillshade-shadow-color":
                hillshadeLayer.hillshadeShadowColor = expression
            case "hillshade-highlight-color":
                hillshadeLayer.hillshadeHighlightColor = expression
            case "hillshade-accent-color":
                hillshadeLayer.hillshadeAccentColor = expression
            case "visibility":
                hillshadeLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    class func addHeatmapProperties(
        heatmapLayer: MGLHeatmapStyleLayer,
        properties: [String: String]
    ) {
        for (propertyName, propertyValue) in properties {
            let expression = interpretExpression(
                propertyName: propertyName,
                expression: propertyValue
            )
            switch propertyName {
            case "heatmap-radius":
                heatmapLayer.heatmapRadius = expression
            case "heatmap-weight":
                heatmapLayer.heatmapWeight = expression
            case "heatmap-intensity":
                heatmapLayer.heatmapIntensity = expression
            case "heatmap-color":
                heatmapLayer.heatmapColor = expression
            case "heatmap-opacity":
                heatmapLayer.heatmapOpacity = expression
            case "visibility":
                heatmapLayer.isVisible = propertyValue == "visible"

            default:
                break
            }
        }
    }

    private class func interpretExpression(propertyName: String,
                                           expression: String) -> NSExpression?
    {
        let isColor = propertyName.contains("color")

        do {
            let json = try JSONSerialization.jsonObject(
                with: expression.data(using: .utf8)!,
                options: .fragmentsAllowed
            )
            // this is required because NSExpression.init(mglJSONObject: json) fails to create
            // a proper Expression if the data of is a hexString
            if isColor {
                if let color = json as? String {
                    return NSExpression(forConstantValue: UIColor(hexString: color))
                }
            }
            // this is required because NSExpression.init(mglJSONObject: json) fails to create
            // a proper Expression if the data of a literal is an array
            if let offset = json as? [Any] {
                if offset.count == 2, offset.first is String, offset.first as? String == "literal" {
                    if let vector = offset.last as? [Any] {
                        if vector.count == 2 {
                            if let x = vector.first as? Double, let y = vector.last as? Double {
                                return NSExpression(
                                    forConstantValue: NSValue(cgVector: CGVector(dx: x,
                                                                                 dy: y))
                                )
                            }
                        }
                    }
                }
            }
            return NSExpression(mglJSONObject: json)
        } catch {}
        return nil
    }
}
